xwiki-platform-core/xwiki-platform-web/src/main/webapp/templates/adduorg.vm

```diff
@@ -20,19 +20,36 @@
 #set ($isAjaxRequest = $request.getHeader('X-Requested-With') == 'XMLHttpRequest')
 #if ($services.csrf.isTokenValid($request.form_token))
   #set ($dirty = false)
+  #set ($ignoredGroups = [])
   #foreach ($name in $request.getParameterValues('name'))
     ## Verify if the user or group is already in the target group.
-    #if ($name.trim() != '' && !$doc.getObject('XWiki.XWikiGroups', 'member', $name))
+    #if ($name.trim() == '')
+      #set ($accept = false)
+    #else
+      #set ($accept = $services.user.group.canAddAsMember($name, $doc.getDocumentReference()))
```

Comment:
since canAddAsMember could throw exception I'd protect it, see: https://extensions.xwiki.org/xwiki/bin/view/Extension/Velocity%20Module#HVelocityDirectives

Reply:
Instead I have captured the exception inside canAddAsMember (see https://github.com/xwiki/xwiki-platform/pull/1374/files#diff-8dc9d53466040f00f533fc828198b1dfR176). WDYT ?

Reply:
Sounds ok to me, now I still have a question: this API is supposed to be called whenever you add group to another group, or whenever you add a member (user or group) in another group. In other word, is that possible that canAddAsMember will be called with $name being a reference to a user instead of a group and in that case how does it behave, in particular would the call to `getMembers(candidate, true)` throw an exception?

Reply:
> in particular would the call to getMembers(candidate, true) throw an exception

AFAIK it should simply return an empty list because no XWiki.XWikiGroups object is found under this reference.

Reply:
Ok was expecting this but wanted to be sure.

Reply:
You were right to ask the question @surli , when trying to add a use that is already a member, the warning message is `Members successfully added but some groups have been ignored (XWiki.U2)`. I'll revert it back to the current message (`Members successfully added`).